src/state/cached_state.rs

```diff
@@ -126,16 +124,34 @@ impl<T: StateReader> StateReader for CachedState<T> {
         }
 
         // I: FETCHING FROM CACHE
-        if let Some(compiled_class) = self.contract_classes.get(class_hash) {
+        let mut private_cache = self.contract_class_cache_private.borrow_mut();
+        if let Some(compiled_class) = private_cache.get(class_hash) {
             return Ok(compiled_class.clone());
+        } else if let Some(compiled_class) =
+            self.contract_class_cache().get_contract_class(*class_hash)
+        {
+            private_cache.insert(*class_hash, compiled_class.clone());
+            return Ok(compiled_class);
         }
 
         // I: CASM CONTRACT CLASS : CLASS_HASH
         if let Some(compiled_class_hash) =
             self.cache.class_hash_to_compiled_class_hash.get(class_hash)
         {
-            if let Some(casm_class) = self.contract_classes.get(compiled_class_hash) {
+            if let Some(casm_class) = self
+                .contract_class_cache_private
+                .borrow()
+                .get(compiled_class_hash)
```

Comment:
Can't we reuse the previous borrow?
```suggestion
            if let Some(casm_class) = private_cache.get(compiled_class_hash)
```

Reply:
Fixed.